kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableSet/TrieNode.kt

```diff
@@ -20,22 +20,46 @@ package kotlinx.collections.immutable.implementations.immutableSet
 internal const val MAX_BRANCHING_FACTOR = 32
 internal const val LOG_MAX_BRANCHING_FACTOR = 5
 internal const val MAX_BRANCHING_FACTOR_MINUS_ONE = MAX_BRANCHING_FACTOR - 1
-internal const val ENTRY_SIZE = 2
+internal const val ENTRY_SIZE = 1
 internal const val MAX_SHIFT = 30
 
+/**
+ * Gets trie index segment of the specified [index] at the level specified by [shift].
+ *
+ * `shift` equal to zero corresponds to the root level.
+ * For each lower level `shift` increments by [LOG_MAX_BRANCHING_FACTOR].
+ */
+internal fun indexSegment(index: Int, shift: Int): Int =
+        (index shr shift) and MAX_BRANCHING_FACTOR_MINUS_ONE
+
+
+private fun <E> Array<Any?>.addElementAtIndex(index: Int, element: E): Array<Any?> {
+    val newBuffer = arrayOfNulls<Any?>(this.size + 1)
+    this.copyInto(newBuffer, endIndex = index)
+    this.copyInto(newBuffer, index + 1, index, this.size)
+    newBuffer[index] = element
+    return newBuffer
+}
+
+private fun Array<Any?>.removeCellAtIndex(cellIndex: Int): Array<Any?> {
+    val newBuffer = arrayOfNulls<Any?>(this.size - 1)
+    this.copyInto(newBuffer, endIndex = cellIndex)
+    this.copyInto(newBuffer, cellIndex, cellIndex + 1, this.size)
+    return newBuffer
+}
 
 internal class TrieNode<E>(var bitmap: Int,
                            var buffer: Array<Any?>,
                            var marker: Marker?) {
 
     constructor(bitmap: Int, buffer: Array<Any?>) : this(bitmap, buffer, null)
 
-    private fun isNullCellAt(position: Int): Boolean {
-        return bitmap and position == 0
+    private fun hasCellAt(positionMask: Int): Boolean {
```

Comment:
Should be negative of "hasCell", or body of this function should be modified to reflect function's name.

Reply:
Good catch